File: leetcode/40-combination-sum-ii.py
# https://leetcode.com/problems/combination-sum-ii/

from typing import List

class Solution:
	def combinationSum2(self, candidates: List[int], target: int) -> List[List[int]]:
		candidates.sort()
		combinations = list()
		combination = list()

		def backtracking(idx, curSum):
			if curSum > target:
				return
			elif curSum == target:
				combinations.append(combination.copy())
				return

			for i in range(idx, len(candidates)):
				if i > idx and candidates[i-1] == candidates[i]:
					continue
				combination.append(candidates[i])
				backtracking(i+1, combination, curSum + candidates[i])
				combination.pop()

		backtracking(0, 0)
		return combinations

File: leetcode/40-combination-sum-ii.js
// https://leetcode.com/problems/combination-sum-ii/

/**
 * @param {number[]} candidates
 * @param {number} target
 * @return {number[][]}
 */
var combinationSum2 = function(candidates, target) {
  const combinations = [];
  const combination = [];

  candidates.sort((a, b) => (a - b));

  const backtracking = (idx = 0, sum = 0) => {

    if (sum > target) {
      return;
    }

    if (sum === target) {
      combinations.push([...combination]);
      return;
    }

    for (let i = idx; i < candidates.length; i++) {
      // the trick is that at with in the same lavel the values should be unique...
      if (i > idx && candidates[i-1] === candidates[i]) continue;
      
      combination.push(candidates[i]);
      backtracking(i+1, sum + candidates[i]);
      combination.pop();
    }
  }

  backtracking();
  return combinations;
};